File: app/src/main/java/io/github/wulkanowy/ui/main/MainActivity.java
package io.github.wulkanowy.ui.main;

import android.content.Context;
import android.content.Intent;
import android.graphics.Color;
import android.os.Bundle;

import com.aurelhubert.ahbottomnavigation.AHBottomNavigation;
import com.aurelhubert.ahbottomnavigation.AHBottomNavigationItem;
import com.aurelhubert.ahbottomnavigation.AHBottomNavigationViewPager;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.github.wulkanowy.R;
import io.github.wulkanowy.ui.base.BaseActivity;
import io.github.wulkanowy.ui.base.BaseFragment;
import io.github.wulkanowy.ui.main.dashboard.DashboardFragment;
import io.github.wulkanowy.ui.main.grades.GradesFragment;
import io.github.wulkanowy.ui.main.timetable.TimetableFragment;

public class MainActivity extends BaseActivity implements MainContract.View,
        AHBottomNavigation.OnTabSelectedListener {

    private static final int DEFAULT_TAB_POSITION = 0;

    @BindView(R.id.main_activity_nav)
    AHBottomNavigation bottomNavigation;

    @BindView(R.id.main_activity_view_pager)
    AHBottomNavigationViewPager viewPager;

    @Inject
    MainPagerAdapter pagerAdapter;

    @Inject
    MainContract.Presenter presenter;

    public static Intent getStartIntent(Context context) {
        return new Intent(context, MainActivity.class);
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        getActivityComponent().inject(this);
        setButterKnife(ButterKnife.bind(this));

        presenter.onStart(this);

        setUpOnCreate();
    }

    @Override
    protected void setUpOnCreate() {
        initiationViewPager();
        initiationBottomNav();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        presenter.onDestroy();
    }

    @Override
    public boolean onTabSelected(int position, boolean wasSelected) {
        presenter.onTabSelected(position, wasSelected, DEFAULT_TAB_POSITION);
        return true;
    }

    @Override
    public void setCurrentPage(int position) {
        viewPager.setCurrentItem(position, false);
    }

    @Override
    public void setChildFragmentSelected(int position, boolean selected) {
        ((BaseFragment) pagerAdapter.getItem(position)).setSelected(selected);
    }

    private void initiationBottomNav() {
        bottomNavigation.addItem(new AHBottomNavigationItem(
                getString(R.string.grades_text),
                getResources().getDrawable(R.drawable.icon_grade_26dp)
        ));
        bottomNavigation.addItem(new AHBottomNavigationItem(
                getString(R.string.attendance_text),
                getResources().getDrawable(R.drawable.icon_attendance_24dp)
        ));
        bottomNavigation.addItem(new AHBottomNavigationItem(
                getString(R.string.dashboard_text),
                getResources().getDrawable(R.drawable.ic_dashboard_black_24dp)
        ));
        bottomNavigation.addItem(new AHBottomNavigationItem(
                getString(R.string.lessonplan_text),
                getResources().getDrawable(R.drawable.icon_lessonplan_24dp)
        ));
        bottomNavigation.addItem(new AHBottomNavigationItem(
                getString(R.string.settings_text),
                getResources().getDrawable(R.drawable.icon_other_24dp)
        ));

        bottomNavigation.setAccentColor(getResources().getColor(R.color.colorPrimary));
        bottomNavigation.setInactiveColor(Color.BLACK);
        bottomNavigation.setBackgroundColor(getResources().getColor(R.color.colorBackgroundBottomNavi));
        bottomNavigation.setTitleState(AHBottomNavigation.TitleState.ALWAYS_SHOW);
        bottomNavigation.setOnTabSelectedListener(this);
        bottomNavigation.setCurrentItem(DEFAULT_TAB_POSITION);
    }

    private void initiationViewPager() {
        pagerAdapter.addFragment(new GradesFragment());
        pagerAdapter.addFragment(new DashboardFragment());
        pagerAdapter.addFragment(new DashboardFragment());
        pagerAdapter.addFragment(new TimetableFragment());
        pagerAdapter.addFragment(new DashboardFragment());

        viewPager.setPagingEnabled(false);
        viewPager.setAdapter(pagerAdapter);
        viewPager.setOffscreenPageLimit(4);
        viewPager.setCurrentItem(DEFAULT_TAB_POSITION, false);
    }
}

File: app/src/main/java/io/github/wulkanowy/ui/main/MainActivity.kt
package io.github.wulkanowy.ui.main

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.content.ContextCompat
import com.aurelhubert.ahbottomnavigation.AHBottomNavigation
import com.aurelhubert.ahbottomnavigation.AHBottomNavigationItem
import com.ncapdevi.fragnav.FragNavController
import com.ncapdevi.fragnav.FragNavController.Companion.DETACH_ON_NAVIGATE_HIDE_ON_SWITCH
import io.github.wulkanowy.R
import io.github.wulkanowy.ui.base.BaseActivity
import io.github.wulkanowy.ui.main.attendance.AttendanceFragment
import io.github.wulkanowy.ui.main.exam.ExamFragment
import io.github.wulkanowy.ui.main.grade.GradeFragment
import io.github.wulkanowy.ui.main.more.MoreFragment
import io.github.wulkanowy.ui.main.timetable.TimetableFragment
import kotlinx.android.synthetic.main.activity_main.*
import javax.inject.Inject

class MainActivity : BaseActivity(), MainView, FragNavController.TransactionListener {
    @Inject
    lateinit var presenter: MainPresenter

    @Inject
    lateinit var navController: FragNavController

    companion object {
        fun getStartIntent(context: Context) = Intent(context, MainActivity::class.java)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        messageView = mainContainer
        presenter.attachView(this)
        navController.initialize(0, savedInstanceState)
    }

    override fun initFragmentController() {
        navController.run {
            rootFragments = listOf(
                    GradeFragment.newInstance(),
                    AttendanceFragment.newInstance(),
                    ExamFragment.newInstance(),
                    TimetableFragment.newInstance(),
                    MoreFragment.newInstance()
            )
            fragmentHideStrategy = DETACH_ON_NAVIGATE_HIDE_ON_SWITCH
            createEager = true
            transactionListener = this@MainActivity
        }
    }

    override fun initBottomNav() {
        mainBottomNav.run {
            addItems(mutableListOf(
                    AHBottomNavigationItem(R.string.grades_text, R.drawable.ic_menu_grade_26dp, 0),
                    AHBottomNavigationItem(R.string.attendance_text, R.drawable.ic_menu_attendance_24dp, 0),
                    AHBottomNavigationItem(R.string.exams_text, R.drawable.ic_menu_exams_24dp, 0),
                    AHBottomNavigationItem(R.string.timetable_text, R.drawable.ic_menu_timetable_24dp, 0),
                    AHBottomNavigationItem(R.string.more_text, R.drawable.ic_menu_other_24dp, 0)
            ))
            accentColor = ContextCompat.getColor(context, R.color.colorPrimary)
            inactiveColor = ContextCompat.getColor(context, android.R.color.black)
            titleState = AHBottomNavigation.TitleState.ALWAYS_SHOW
            currentItem = 0
            isBehaviorTranslationEnabled = false
            setOnTabSelectedListener { position, _ ->
                navController.switchTab(position)
                true
            }
        }
    }

    override fun onFragmentTransaction(fragment: Fragment?, transactionType: FragNavController.TransactionType) {

    }

    override fun onTabTransaction(fragment: Fragment?, index: Int) {
        setTitle(when (index) {
            0 -> R.string.grades_text
            1 -> R.string.attendance_text
            2 -> R.string.exams_text
            3 -> R.string.timetable_text
            4 -> R.string.more_text
            else -> R.string.activity_dashboard_text //TODO Change default title
        })
    }

    override fun onSaveInstanceState(outState: Bundle?) {
        super.onSaveInstanceState(outState)
        navController.onSaveInstanceState(outState)
    }
}

